Add filter option to ol.format.WFS#writeGetFeature

tsauerwein opened the pull request:
[ol.format.WFS#writeGetFeature](http://openlayers.org/en/master/apidoc/ol.format.WFS.html#writeGetFeature) currently only allows to set a bbox filter, this PR proposes to add a `filter` option which supports a sub-set of the OGC filters.

Example:
```javascript
var request = new ol.format.WFS().writeGetFeature({
  ...
  filter: ol.format.wfs.filter.equalTo('name', 'New York')
});
```
Or:
```javascript
var f = ol.format.wfs.filter;
var request = new ol.format.WFS().writeGetFeature({
  ...
  filter: f.or(f.equalTo('name', 'New York'), f.equalTo('area', 1234))
});
```

The following filters are currently supported:

* Not/And/Or
* PropertyIsEqualTo/PropertyIsNotEqualTo
* PropertyIsLessThan/PropertyIsLessThanOrEqualTo
* PropertyIsGreaterThan/PropertyIsGreaterThanOrEqualTo
* PropertyIsLike
* PropertyIsNull
* PropertyIsBetween
* BBOX

tsauerwein commented:
Anyone up for a review? @bartvde maybe, our expert for OGC topics? :)

bartvde commented:
@tsauerwein I can have a look at this but won't have time until later in this week (mid/end).

tsauerwein commented:
Thanks, Bart! @fredj will take a look, but a 2nd opinion is welcome!

fredj commented on src/ol/format/wfs/filter.js:
```diff
@@ -0,0 +1,617 @@
+ * @extends {ol.format.wfs.filter.LogicalBinary}
+ * @api
+ */
+ol.format.wfs.filter.And = function(conditionA, conditionB) {
```
Is it really needed to implement this function if it only calls the parent? if it for jsdoc ?

fredj replied:
forget the last comment, slow back from vacation ...

fredj commented:
LGTM but I would like to have a review from Bart

bartvde commented:
I'll have a quick look now.

bartvde commented on src/ol/format/wfs/filter.js:
```diff
@@ -0,0 +1,617 @@
+goog.provide('ol.format.wfs.filter');
```
shouldn't we put this in ```ol.format.ogc.filter``` instead? Since filter can be used outside of wfs and is part of the ogc namespace not the wfs one.

tsauerwein replied:
Yes, that makes sense.

bartvde commented on src/ol/format/wfs/filter.js:
```diff
@@ -0,0 +1,617 @@
+
+
+/**
+ * Creates a `<PropertyIsBetween>` comparison operator to  test whether an expression
```
two spaces after to

bartvde commented on src/ol/format/wfsformat.js:
```diff
@@ -551,43 +559,179 @@ ol.format.WFS.writeQuery_ = function(node, featureType, objectStack) {
+  node.setAttribute('wildCard', filter.wildCard);
+  node.setAttribute('singleChar', filter.singleChar);
+  node.setAttribute('escapeChar', filter.escapeChar);
+  ol.format.WFS.writeOgcPropertyName_(node, filter.propertyName);
```
minor but missing support for matchCase here

tsauerwein replied:
If I get the OGC specification right, `matchCase` is not supposed to be used with `PropertyIsLike`, only with `PropertyIsEqualTo` and `PropertyIsNotEqualTo`. Correct?

bartvde replied:
I've always thought that was a mistake in the schema, and I even reported it as early as 2006 I think: http://www.ogcnetwork.net/node/962
https://trac.osgeo.org/openlayers/ticket/3247
http://osgeo-org.1560.x6.nabble.com/ogc-filter-1-1-doesn-t-support-matchCase-in-PropertyIsLike-td5079577.html

tsauerwein replied:
Ok, thanks for the info. I will also add it to `PropertyIsLike`.

bartvde commented:
Nice work @tsauerwein I only put a few minor remarks in there. Otherwise LGTM.

tsauerwein commented:
Thanks a lot for the reviews! I renamed `ol.format.wfs.filter` to `ol.format.ogc.filter` and added support for the `matchCase` attribute. I will merge when Travis agrees.

bartvde commented:
thanks @tsauerwein !

tschaub commented on src/ol/format/ogc/filter.js:
```diff
@@ -0,0 +1,637 @@
+   */
+  this.tagName_ = tagName;
+};
+goog.inherits(ol.format.ogc.filter.Filter, ol.Object);
```
I'm curious why this inherits from `ol.Object`.  Is this used anywhere?

tsauerwein replied:
Good point, there should be no need to inherit from `ol.Object`.

fredj replied:
I'm creating a pull request to fix it
